drivers/SmartThings/matter-lock/src/init.lua

```diff
@@ -29,59 +29,51 @@ local lock_utils = require "lock_utils"
     -- Shouldn't happen but defensive to try to figure out if we need the cota cred and set it.
     device:send(DoorLock.attributes.RequirePINforRemoteOperation:read(device, #eps > 0 and eps[1] or 1))
-    device.thread:call_with_delay(2, function(t) set_cota_credential(device) end)
+    device.thread:call_with_delay(2, function(t) set_cota_credential(device, credential_index) end)
   elseif not cota_cred then
     device.log.debug("Device does not require PIN for remote operation. Not setting COTA credential")
     return
   end
 
-  if device:get_latest_state(
-    "main", capabilities.lockCodes.ID, capabilities.lockCodes.scanCodes.NAME
-  ) == "Scanning" then
+  if device:get_field(lock_utils.SET_CREDENTIAL) ~= nil then
+    device.log.debug("delaying setting COTA credential since a credential is currently being set")
     device.thread:call_with_delay(2, function(t)
-      set_cota_credential(device)
+      set_cota_credential(device, credential_index)
     end)
     return
   end
 
-  --try to use last code slot in hopes that it wont overwrite existing codes on the device
-  local credential_index = device:get_field(lock_utils.TOTAL_PIN_USERS) or
-    device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.maxCodes.NAME) or 1
+  device:set_field(lock_utils.COTA_CRED_INDEX, credential_index, {persist = true})
   local credential = {credential_type = DoorLock.types.DlCredentialType.PIN, credential_index = credential_index}
-
-  -- Clear the credential to make sure that we have an open slot for the cota credential
-  device.thread:call_with_delay(0, function(t)
-    --Note we dont set lock_utils.DELETEING_CODE field to avoid re-setting cota credential this time
-    device:send(DoorLock.server.commands.ClearCredential(
-      device,
-      #eps > 0 and eps[1] or 1,
-      credential
-    ))
-  end)
-
   -- Set the credential to a code
-  device.thread:call_with_delay(2, function(t)
-    device:set_field(lock_utils.SET_CREDENTIAL, credential_index)
-    device:send(DoorLock.server.commands.SetCredential(
-      device,
-      #eps > 0 and eps[1] or 1,
-      DoorLock.types.DlDataOperationType.ADD,
-      credential,
-      device:get_field(lock_utils.COTA_CRED),
-      nil, -- nil user_index creates a new user
-      DoorLock.types.DlUserStatus.OCCUPIED_ENABLED,
-      DoorLock.types.DlUserType.REMOTE_ONLY_USER
-    ))
-  end)
+  device:set_field(lock_utils.SET_CREDENTIAL, credential_index)
+  device.log.info(string.format("Attempting to set COTA credential at index %s", credential_index))
+  device:send(DoorLock.server.commands.SetCredential(
+    device,
+    #eps > 0 and eps[1] or 1,
+    DoorLock.types.DlDataOperationType.ADD,
+    credential,
+    device:get_field(lock_utils.COTA_CRED),
+    nil, -- nil user_index creates a new user
+    DoorLock.types.DlUserStatus.OCCUPIED_ENABLED,
+    DoorLock.types.DlUserType.REMOTE_ONLY_USER
+  ))
+end
+
+local function generate_cota_cred_for_device(device)
```

Comment:
should this be a no-op if there's already a generated credential?

Reply:
No. There is a case, although unlikely, where our generated credential is the same as a PIN that already exists on the lock. In this case we fail to set the credential, and need to regenerate, or else we would continue to fail to set the credential.

Reply:
> There is a case, although unlikely, where our generated credential is the same as a PIN that already exists on the lock

What are the security implications here? Checking to see if a code is already in use and then exhibiting specific behaviors on collision is an actionable form of brute force password sniffing in the general case. I don't know enough about Matter or what else is going on on the system to know if this is actionable in this specific case but I figure it was worth asking.

Reply:
This came up in my conversation with Yale. There is a limit to the number of wrong codes that can be entered manually (WrongCodeEntryLimit) which does apply to PINs in remote commands. This mitigates one command that can be used to brute force attack, but there is still the SetCredential to discover duplicates in a brute force way... 

<img width="1048" alt="image" src="https://user-images.githubusercontent.com/20526748/212979929-80afc242-2cf4-45d3-9492-8324f30298f4.png">

Reply:
The device would need to be commissioned to a malicious fabric though, which requires user interaction to commission the device.